File: python/src/it_works_on_my_machine/probability.py
"""
Probability utilities for flaky tests.
"""

import secrets


def random_success(success_rate: float) -> bool:
    """
    Return True with the given success rate (0.0 to 1.0).
    
    Args:
        success_rate: Probability of success (0.0 = never, 1.0 = always)
        
    Returns:
        True if random number is below success rate, False otherwise
    """
    if not 0.0 <= success_rate <= 1.0:
        raise ValueError("Success rate must be between 0.0 and 1.0")
    
    random_number = secrets.randbelow(100)
    return random_number < (success_rate * 100)


def dice_roll() -> int:
    """
    Simulate rolling a six-sided die.
    
    Returns:
        Random number between 1 and 6 (inclusive)
    """
    return secrets.randbelow(6) + 1


def is_baseline_test() -> bool:
    """
    Always returns True - for baseline reliability tests.
    
    Returns:
        Always True
    """
    return True

File: typescript/src/probability.ts
/**
 * Probability utilities for flaky tests.
 */

/**
 * Return true with the given success rate (0.0 to 1.0).
 * 
 * @param successRate - Probability of success (0.0 = never, 1.0 = always)
 * @returns True if random number is below success rate, false otherwise
 */
export function randomSuccess(successRate: number): boolean {
  if (successRate < 0.0 || successRate > 1.0) {
    throw new Error('Success rate must be between 0.0 and 1.0');
  }
  
  const randomNumber = Math.floor(Math.random() * 100);
  return randomNumber < (successRate * 100);
}

/**
 * Simulate rolling a six-sided die.
 * 
 * @returns Random number between 1 and 6 (inclusive)
 */
export function diceRoll(): number {
  return Math.floor(Math.random() * 6) + 1;
}

/**
 * Always returns true - for baseline reliability tests.
 * 
 * @returns Always true
 */
export function isBaselineTest(): boolean {
  return true;
}

/**
 * Calculate the probability of all tests passing given individual success rates.
 * 
 * @param successRates - Array of individual test success rates
 * @returns Overall probability of all tests passing
 */
export function calculateOverallSuccessRate(successRates: number[]): number {
  return successRates.reduce((overall, rate) => overall * rate, 1.0);
}
